### PR TITLE
Civilian

### PR DESCRIPTION
**Pull request by Shelbyen**

### CODE REVIEW

File: src/main/resources/assets/worldbox/models/item/civilian_spawn_egg.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "parent": "item/template_spawn_egg"
+}
```
Comment (jaka2005) [inline src/main/resources/assets/worldbox/models/item/civilian_spawn_egg.json]: No new line at end of file

File: src/main/java/com/j2k/worldbox/item/custom/ModSpawnEggItem.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import net.minecraftforge.fml.RegistryObject;
+import net.minecraftforge.fml.common.ObfuscationReflectionHelper;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class ModSpawnEggItem extends SpawnEggItem {
+    protected static final List<ModSpawnEggItem> UNADDED_EGGS = new ArrayList<>();
+    private final Lazy<? extends EntityType<?>> entityTypeSupplier;
```
Comment (jaka2005) [inline src/main/java/com/j2k/worldbox/item/custom/ModSpawnEggItem.java]: why do you use lazy initialization?

File: src/main/java/com/j2k/worldbox/item/ModItems.java [hunk 3]
```diff
@@ -11,6 +14,10 @@ public class ModItems {
     public static final DeferredRegister<Item> ITEMS =
             DeferredRegister.create(ForgeRegistries.ITEMS, WorldBoxMod.MOD_ID);
 
+    public static final RegistryObject<ModSpawnEggItem> CIVILIAN_SPAWN_EGG = ITEMS.register("civilian_spawn_egg",
+            () -> new ModSpawnEggItem(ModEntityTypes.CIVILIAN, 0xFFC8A8, 0xFF4F00,
```
Comment (jaka2005) [inline src/main/java/com/j2k/worldbox/item/ModItems.java]: what are these magic numbers?

File: src/main/java/com/j2k/worldbox/entity/render/CivilianRender.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import com.j2k.worldbox.entity.custom.CivilianEntity;
+import com.j2k.worldbox.entity.model.CivilianModel;
+import net.minecraft.client.renderer.entity.EntityRendererManager;
+import net.minecraft.client.renderer.entity.MobRenderer;
+import net.minecraft.util.ResourceLocation;
+
+public class CivilianRender extends MobRenderer<CivilianEntity, CivilianModel<CivilianEntity>> {
+    protected static final ResourceLocation TEXTURE =
+            new ResourceLocation(WorldBoxMod.MOD_ID, "textures/entity/civilian.png");
+
```
Comment (jaka2005) [inline src/main/java/com/j2k/worldbox/entity/render/CivilianRender.java]: write comments explaining the meanings of magic numbers

File: src/main/java/com/j2k/worldbox/entity/model/CivilianModel.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import com.mojang.blaze3d.matrix.MatrixStack;
+import com.mojang.blaze3d.vertex.IVertexBuilder;
+import net.minecraft.client.renderer.entity.model.BipedModel;
+import net.minecraft.client.renderer.model.ModelRenderer;
+
+public class CivilianModel<T extends CivilianEntity> extends BipedModel<T> {
+
+	protected CivilianModel(float modelSize, float yOffsetIn, int textureWidthIn, int textureHeightIn) {
+		super(modelSize, yOffsetIn, textureWidthIn, textureHeightIn);
+		// NOTE:XXX: There may be problems with rotation points
```
Comment (jaka2005) [inline src/main/java/com/j2k/worldbox/entity/model/CivilianModel.java]: can there still be problems here?

File: src/main/java/com/j2k/worldbox/entity/custom/CivilianEntity.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+public class CivilianEntity extends MobEntity {
+
+    private final double rating;
+    private boolean freedom;
+    private static int ticks = 0;
+
+    public CivilianEntity(EntityType<? extends MobEntity> type, World worldIn) {
+        super(type, worldIn);
+        this.rating = ThreadLocalRandom.current().nextDouble();
```
Comment (jaka2005) [inline src/main/java/com/j2k/worldbox/entity/custom/CivilianEntity.java]: this is cool!

File: src/main/java/com/j2k/worldbox/buildings/Building.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+    public ArrayList<CivilianEntity> getBuilders() {
+        return new ArrayList<>(this.builders);
+    }
+
+    public void addBuilders(CivilianEntity newBuilder) {
+        this.builders.add(newBuilder);
+    }
+
+    protected ArrayList<CivilianEntity> removeBuilders() {
```
Comment (jaka2005) [inline src/main/java/com/j2k/worldbox/buildings/Building.java]: it should not return an ArrayList
it can return Boolean

Comment (jaka2005) [inline src/main/java/com/j2k/worldbox/buildings/Building.java]: why it protected?

File: src/main/java/com/j2k/worldbox/buildings/Building.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
+package com.j2k.worldbox.buildings;
+
+import com.j2k.worldbox.entity.custom.CivilianEntity;
+import net.minecraft.util.math.BlockPos;
+
+import java.util.ArrayList;
+
+public abstract class Building {
```
Comment (jaka2005) [inline src/main/java/com/j2k/worldbox/buildings/Building.java]: you can use abstract attributes (abstract fields and abstract methods)

**Review by jaka2005 — CHANGES_REQUESTED** [review]
write more comments explaining what is happening in the code and learn basics of OOP;
please rename domains from "com" to "site"

File: src/main/java/com/j2k/worldbox/item/ModItems.java (hunk 3 above)
Comment (Shelbyen) [reply src/main/java/com/j2k/worldbox/item/ModItems.java]: colors

File: src/main/java/com/j2k/worldbox/entity/render/CivilianRender.java (hunk 4 above)
Comment (Shelbyen) [reply src/main/java/com/j2k/worldbox/entity/render/CivilianRender.java]: where are the magic values used here

File: src/main/java/com/j2k/worldbox/entity/model/CivilianModel.java (hunk 5 above)
Comment (Shelbyen) [reply src/main/java/com/j2k/worldbox/entity/model/CivilianModel.java]: I didn't force them to walk, so most likely there may be problems

File: src/main/java/com/j2k/worldbox/entity/custom/CivilianEntity.java (hunk 6 above)
Comment (Shelbyen) [reply src/main/java/com/j2k/worldbox/entity/custom/CivilianEntity.java]: thx

File: src/main/java/com/j2k/worldbox/item/ModItems.java (hunk 3 above)
Comment (jaka2005) [reply src/main/java/com/j2k/worldbox/item/ModItems.java]: colors of what?

File: src/main/java/com/j2k/worldbox/entity/render/CivilianRender.java (hunk 4 above)
Comment (jaka2005) [reply src/main/java/com/j2k/worldbox/entity/render/CivilianRender.java]: sorry, magic values on line 15 
```java
super(renderManagerIn, new CivilianModel<>(1F), 0.7F);
```

File: src/main/java/com/j2k/worldbox/entity/model/CivilianModel.java (hunk 5 above)
Comment (jaka2005) [reply src/main/java/com/j2k/worldbox/entity/model/CivilianModel.java]: okey, thx

File: src/main/java/com/j2k/worldbox/item/ModItems.java (hunk 3 above)
Comment (Shelbyen) [reply src/main/java/com/j2k/worldbox/item/ModItems.java]: Hmmm... Maybe for mob spawn egg?... Hmmmm

**Review by jaka2005 — APPROVED** [review]
thx for contribution!!

**Pull request merged**